lib/src/config_concatenation.cc

```diff
@@ -0,0 +1,248 @@
+#include <hocon/config_object.hpp>
+#include <hocon/path.hpp>
+#include <internal/config_concatenation.hpp>
+#include <internal/values/config_string.hpp>
+#include <internal/default_transformer.hpp>
+#include <internal/simple_config_list.hpp>
+#include <internal/simple_config_origin.hpp>
+#include <internal/config_exception.hpp>
+#include <internal/values/config_string.hpp>
+#include <internal/resolve_result.hpp>
+#include <internal/resolve_source.hpp>
+#include <internal/resolve_context.hpp>
+
+using namespace std;
+
+namespace hocon {
+
+    config_concatenation::config_concatenation(shared_origin origin, std::vector<shared_value> pieces) :
+            config_value(move(origin)), _pieces(move(pieces)), _unmerged_values({ shared_from_this() }) {
+        if (_pieces.size() < 2) {
+            throw config_exception("Created concatenation with less than 2 items");
+        }
+
+        bool had_unmergeable = false;
+        for (shared_value p : _pieces) {
+            if (dynamic_pointer_cast<const config_concatenation>(p)) {
+                throw config_exception("config_concatenation should never be nested");
+            }
+
+            if (dynamic_pointer_cast<const unmergeable>(p)) {
+                had_unmergeable = true;
+            }
+        }
+
+        if (!had_unmergeable) {
+            throw config_exception("Created concatenation without an unmergeable in it");
+        }
+    }
+
+    vector<shared_value> config_concatenation::pieces() const {
+        return _pieces;
+    }
+
+    config_value_type config_concatenation::value_type() const {
+        throw not_resolved();
+    }
+
+    vector<shared_value> const& config_concatenation::unmerged_values() const {
+        return _unmerged_values;
+    }
+
+    resolve_status config_concatenation::get_resolve_status() const {
+        return resolve_status::UNRESOLVED;
+    }
+
+    shared_value config_concatenation::replace_child(shared_value const& child, shared_value replacement) const {
+        auto new_list = config_value::replace_child_in_list(_pieces, child, replacement);
+        if (new_list.empty()) {
+            return nullptr;
+        } else {
+            return make_shared<config_concatenation>(origin(), new_list);
+        }
+    }
+
+    bool config_concatenation::has_descendant(shared_value const &descendant) const {
+        return has_descendant_in_list(_pieces, descendant);
+    }
+
+    resolve_result<shared_value>
+    config_concatenation::resolve_substitutions(resolve_context const& context,
+                                                resolve_source const& source) const
+    {
+        // Right now there's no reason to pushParent here because the
+        // content of ConfigConcatenation should not need to replaceChild,
+        // but if it did we'd have to do this.
+        resolve_source source_with_parent = source;
```

Comment:
I think this can just be another const&, no need to make a copy. Which makes me think we can get rid of it and use `context`.